src/display/editor/stamp.js

```diff
@@ -0,0 +1,411 @@
+        this.width * parentWidth,
+        this.height * parentHeight
+      );
+    }
+
+    return this.div;
+  }
+
+  #createCanvas() {
+    const { div } = this;
+    let { width, height } = this.#bitmap;
+    const [pageWidth, pageHeight] = this.pageDimensions;
+    if (this.width) {
+      width = this.width * pageWidth;
+      height = this.height * pageHeight;
+    } else if (width > 0.75 * pageWidth || height > 0.75 * pageHeight) {
+      // If the the image is too big compared to the page dimensions
+      // (more than 75%) then we scale it down.
+      const factor = Math.min(
+        (0.75 * pageWidth) / width,
+        (0.75 * pageHeight) / height
+      );
+      width *= factor;
+      height *= factor;
+    }
+    const [parentWidth, parentHeight] = this.parentDimensions;
+    this.setDims(
+      (width * parentWidth) / pageWidth,
+      (height * parentHeight) / pageHeight
+    );
+
+    this.setAspectRatio(width, height);
+    this.setMinDims();
+
+    const canvas = (this.#canvas = document.createElement("canvas"));
+    div.append(canvas);
+    this.#drawBitmap(width, height);
+    this.#createObserver();
+    div.classList.remove("loading");
+  }
+
+  /**
+   * When the dimensions of the div change the inner canvas must
+   * renew its dimensions, hence it must redraw its own contents.
+   * @param {number} width - the new width of the div
+   * @param {number} height - the new height of the div
+   * @returns
+   */
+  #setDimensions(width, height) {
+    height = this.getHeight(width, height);
+    const [parentWidth, parentHeight] = this.parentDimensions;
+    this.width = width / parentWidth;
+    this.height = height / parentHeight;
+    this.setDims(width, height);
+    if (this.#resizeTimeoutId !== null) {
+      clearTimeout(this.#resizeTimeoutId);
+    }
+    this.#resizeTimeoutId = setTimeout(() => {
+      this.#resizeTimeoutId = null;
+      this.#drawBitmap(width, height);
+    }, 200);
+  }
+
+  #scaleBitmap(width, height) {
+    const { width: bitmapWidth, height: bitmapHeight } = this.#bitmap;
+
+    let newWidth = bitmapWidth;
+    let newHeight = bitmapHeight;
+    let bitmap = this.#bitmap;
+    while (newWidth > 2 * width || newHeight > 2 * height) {
+      const prevWidth = newWidth;
+      const prevHeight = newHeight;
+
+      if (newWidth > 2 * width) {
+        // See bug 1820511 (Windows specific bug).
+        // TODO: once the above bug is fixed we could revert to:
+        // newWidth = Math.ceil(newWidth / 2);
+        newWidth =
+          newWidth >= 16384
+            ? Math.floor(newWidth / 2) - 1
+            : Math.ceil(newWidth / 2);
+      }
+      if (newHeight > 2 * height) {
+        newHeight =
+          newHeight >= 16384
+            ? Math.floor(newHeight / 2) - 1
+            : Math.ceil(newHeight / 2);
+      }
+
+      const offscreenCanvas = new OffscreenCanvas(newWidth, newHeight);
+      const offscreenCtx = offscreenCanvas.getContext("2d");
```

Comment:
Maybe slightly shorter names?
```suggestion
      const offscreen = new OffscreenCanvas(newWidth, newHeight);
      const ctx = offscreenCanvas.getContext("2d");
```